ethereum/src/ERC20Manager.sol

```diff
@@ -0,0 +1,123 @@
+pragma solidity ^0.8.24;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {IERC20Manager} from "./interfaces/IERC20Manager.sol";
+import {IMessageQueueReceiver} from "./interfaces/IMessageQueue.sol";
+import {ERC20GearSupply} from "./ERC20GearSupply.sol";
+import {BridgingPayment} from "./BridgingPayment.sol";
+
+contract ERC20Manager is IERC20Manager, IMessageQueueReceiver {
+    using SafeERC20 for IERC20;
+
+    address immutable MESSAGE_QUEUE_ADDRESS;
+    bytes32 immutable VFT_MANAGER_ADDRESS;
+
+    mapping(address => SupplyType) tokenSupplyType;
+
+    constructor(address message_queue, bytes32 vft_manager) {
+        MESSAGE_QUEUE_ADDRESS = message_queue;
+        VFT_MANAGER_ADDRESS = vft_manager;
+    }
+
+    /** @dev Request token bridging. When the bridging is requested tokens are burned/locked (based on the type of supply)
+     * from account that've sent transaction and `BridgingRequested` event is emitted that later can be verified
+     * on other side of bridge.
+     *
+     * @param token token address to transfer over bridge
+     * @param amount quantity of tokens to transfer over bridge
+     * @param to destination of transfer on gear
+     */
+    function requestBridging(address token, uint256 amount, bytes32 to) public {
+        SupplyType supply_type = tokenSupplyType[token];
+
+        if (supply_type == SupplyType.Gear) {
+            ERC20GearSupply(token).burnFrom(tx.origin, amount);
+        } else {
+            if (supply_type == SupplyType.Unknown) {
+                tokenSupplyType[token] = SupplyType.Ethereum;
+            }
+
+            IERC20(token).safeTransferFrom(tx.origin, address(this), amount);
+        }
+
+        emit BridgingRequested(tx.origin, to, token, amount);
+    }
+
+    /** @dev Accept bridging request made on other side of bridge.
+     * This request must be sent by `MessageQueue` only. When such a request is accepted, tokens
+     * are minted/unlocked to the corresponding account address, specified in `payload`.
+     *
+     * Expected `payload` consisits of these:
```

Comment:
The payload does not contain a nonce? Is it possible that `MESSAGE_QUEUE_ADDRESS` could resend the same payload?